File: tests/comparison/cpp/factorial.cpp
#include <iostream>
#include <cstdlib>

/**
 * Factorial benchmark for C++
 * Usage: ./factorial <n>
 */

long long factorial_recursive(int n) {
    if (n <= 1) {
        return 1;
    }
    return n * factorial_recursive(n - 1);
}

long long factorial_iterative(int n) {
    long long result = 1;
    for (int i = 2; i <= n; ++i) {
        result *= i;
    }
    return result;
}

int main(int argc, char* argv[]) {
    if (argc != 2) {
        std::cerr << "Usage: " << argv[0] << " <n>" << std::endl;
        return 1;
    }
    
    int n = std::atoi(argv[1]);
    if (n < 0) {
        std::cerr << "Error: n must be non-negative" << std::endl;
        return 1;
    }
    
    // Use recursive for reasonable stack depth
    long long result;
    if (n <= 20) {
        result = factorial_recursive(n);
    } else {
        result = factorial_iterative(n);
    }
    
    // Don't print result to avoid affecting timing
    
    return 0;
}

File: tests/comparison/python/factorial.py
#!/usr/bin/env python3
"""
Factorial benchmark for Python
Usage: python factorial.py <n>
"""

import sys

def factorial_recursive(n):
    """Recursive factorial implementation"""
    if n <= 1:
        return 1
    return n * factorial_recursive(n - 1)

def factorial_iterative(n):
    """Iterative factorial implementation - standardized across all languages"""
    result = 1
    i = 2
    while i <= n:
        result *= i
        i += 1
    return result

def main():
    if len(sys.argv) != 2:
        print("Usage: python factorial.py <n>")
        sys.exit(1)
    
    try:
        n = int(sys.argv[1])
    except ValueError:
        print("Error: n must be an integer")
        sys.exit(1)
    
    if n < 0:
        print("Error: n must be non-negative")
        sys.exit(1)
    
    # Use recursive for reasonable stack depth - standardized across all languages
    if n <= 20:
        result = factorial_recursive(n)
    else:
        result = factorial_iterative(n)
    
    # Don't print result to avoid affecting timing

if __name__ == "__main__":
    main()